tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/Selenium/SeleniumCommon.cs

```diff
@@ -0,0 +1,167 @@
+using Datadog.Trace.Logging;
+using Datadog.Trace.Vendors.Newtonsoft.Json;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Testing.Selenium;
+
+internal static class SeleniumCommon
+{
+    internal const string IntegrationName = nameof(IntegrationId.Selenium);
+    private const IntegrationId IntegrationId = Configuration.IntegrationId.Selenium;
+
+    internal const string CommandGet = "get";
+    internal const string CommandClose = "close";
+    internal const string CommandQuit = "quit";
+    private const string CookieName = "datadog-ci-visibility-test-execution-id";
+    private const string RumStopSessionScript = """
+                                                if (window.DD_RUM && window.DD_RUM.stopSession) {
+                                                   window.DD_RUM.stopSession();
+                                                   return true;
+                                                } else {
+                                                   return false;
+                                                }
+                                                """;
+
+    internal static readonly IDatadogLogger Log = CIVisibility.Log;
+    private static Type? _seleniumCookieType;
+    private static long _openPageCount;
+
+    internal static bool IsEnabled => CIVisibility.IsRunning && Tracer.Instance.Settings.IsIntegrationEnabled(IntegrationId);
+
+    internal static void OnBeforePageLoad<TTarget>(TTarget instance, Dictionary<string, object>? parameters)
+        where TTarget : IWebDriverProxy => PreClose(instance);
+
+    internal static void OnPageClose<TTarget>(TTarget instance)
+        where TTarget : IWebDriverProxy => PreClose(instance);
+
+    internal static void OnQuit<TTarget>(TTarget instance)
+        where TTarget : IWebDriverProxy => PreClose(instance);
+
+    private static void PreClose<TTarget>(TTarget instance)
+        where TTarget : IWebDriverProxy
+    {
+        if (Interlocked.Read(ref _openPageCount) > 0)
+        {
+            CloseAndFlush(instance, Test.Current);
+            Interlocked.Decrement(ref _openPageCount);
+        }
+    }
+
+    internal static void OnAfterPageLoad<TTarget>(TTarget instance, Dictionary<string, object>? parameters)
+        where TTarget : IWebDriverProxy
+    {
+        if (Test.Current is not { } test)
+        {
+            return;
+        }
+
+        _seleniumCookieType ??= instance.Type.Assembly.GetType("OpenQA.Selenium.Cookie");
+        if (_seleniumCookieType is not null)
+        {
+            var span = test.GetInternalSpan();
+            var tags = test.GetTags();
+            var traceId = span.Context.TraceId;
+
+            // Create a cookie with the traceId to be used by the RUM library
+            if (Activator.CreateInstance(_seleniumCookieType, CookieName, traceId.ToString()) is { } cookieInstance)
+            {
+                Log.Debug("Inject: {Parameters}", JsonConvert.SerializeObject(parameters ?? new object()));
```

Comment:
Is it really worth creating a new object to be serialized every time this is called? Remember that the args are invoked even if debug is disabled. It's probably worth wrapping in a `Log.IsEnabled(LogLevel.Debug)` either way?
```suggestion
                if (Log.IsEnabled(LogEventLevel.Debug))
                {
                    Log.Debug("Inject: {Parameters}", parameters 
 is null ? null : JsonConvert.SerializeObject(parameters));
                 }
```